### DISCUSSION TITLE
ESP32 - Determine "flash_user_start" from build outputs

### DISCUSSION THREAD

**Discussion by git-n-pissed** [micropython/micropython#10197] [ESP32]
When building for the ESP8266 I'm able to read the ELF output and determine the value that would be returned by `esp.flash_user_start()`, but the ELF file for an ESP32 build does not contain that information.  I've grep'ed the other build outputs for the decimal and hex value I'm expecting (2097152 bytes for an 8MB ESP32-S3), but I'm not finding it.  Is there someway I can glean the flash user start value after a build so I can use it in my build scripts instead of having to hard code a value per board being targeted?

**Comment by jimmo** [comment]
@git-n-pissed On ESP32 the value of flash_user_start is hard coded to exactly that value. See https://github.com/micropython/micropython/blob/master/ports/esp32/modesp.c#L96

**Reply by git-n-pissed** [reply]
Humm, can you shine any light on why that value is what it is?

**Reply by jimmo** [reply]
Because on the original ESP32 port, that's where the user partition is. See partitions.csv -- https://github.com/micropython/micropython/blob/master/ports/esp32/partitions.csv

However, with more boards, flash sizes, and now ESP32 families, the partition layout can change.

I think we should remove this function, in favour of using the `esp32.Partition` class (which is exactly what flashbdev.py does at boot anyway). Partition::info() will let you get the address.

(Just as an example, here's a simple ESP32 that uses the default partition layout, showing that the two agree).

```
>>> import esp
>>> esp.flash_user_start()
2097152
>>> import esp32
>>> p = esp32.Partition("vfs")
>>> p.info()
(1, 129, 2097152, 2097152, 'vfs', False)
```

**Reply by jimmo** [reply]
But to get back to your original question... I don't know how easy it would be to extract the partition table from a built .elf. Look at how the IDF does it here -- https://github.com/espressif/esp-idf/blob/v4.4.3/components/spi_flash/partition.c#L184

But in general, it also depends on other factors -- like with UF2 firmware you don't get to control the partition table. See https://github.com/micropython/micropython/blob/master/ports/esp32/modules/flashbdev.py

**Reply by git-n-pissed** [reply]
> (Just as an example, here's a simple ESP32 that uses the default partition layout, showing that the two agree).
> 
> ```
> >>> import esp
> >>> esp.flash_user_start()
> 2097152
> >>> import esp32
> >>> p = esp32.Partition("vfs")
> >>> p.info()
> (1, 129, 2097152, 2097152, 'vfs', False)
> ```

Yeah, but as you eluded to, that only works with the default partition table.  I guess that also kind of answers my question; I can just look at the partition table CSV.

**Reply by jimmo** [reply]
Not sure I follow what you mean -- it works with whatever the board's partition table is. I was just demonstrating on the board that I had on hand, which just happens to match the 0x20000 value.

**Reply by git-n-pissed** [reply]
> Not sure I follow what you mean -- it works with whatever the board's partition table is. I was just demonstrating on the board that I had on hand, which just happens to match the 0x20000 value.

What I meant is that if `flash_user_start` returns the value hardcoded in modesp.c and I change the offset of `vfs` in the partition table to something other than 0x20000, the two no longer align. Maybe I don't understand what the "user flash space" is. Is it not the starting location for the file system?

**Reply by jimmo** [reply]
Sorry about the delay responding.

I'm not quite sure I follow. What i'm saying is that the `flash_user_start` is incorrect and should be either removed or updated to return the partition start address. The partition start address (defined in partitions.csv) is always correct.
